### PR TITLE
fix: avoid validating checksums for partial responses

### LINKED ISSUE #362: Checksums are validated for partial responses
#### Environment details

  - OS type and version: macOS 12.5.1
  - Python version: 3.10.7
  - pip version: 22.2.2
  - `google-resumable-media` version: 2.4.0

#### Steps to reproduce

This is hard to reproduce as the live GCS doesn't seem to send checksums for partial responses.
This [fake-gcs-server](https://github.com/fsouza/fake-gcs-server) however includes checksums with partial responses (HTTP status 206), which seems to be allowed according to the box [here](https://cloud.google.com/storage/docs/hashes-etags#crc32c).

I would suggest skipping validation when we receive a partial response.

### PR DESCRIPTION
**Pull request by simonbohnen**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-resumable-media-python/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #362 🦕

### CODE REVIEW

**Comment by parthea** [comment]
Hi @simonbohnen,

Please could you sign the CLA?
https://cla.developers.google.com/about

**Comment by simonbohnen** [comment]
> Hi @simonbohnen,
> 
> Please could you sign the CLA? https://cla.developers.google.com/about

It looks like I did for my work account (SimonBohnenQC). Is there a way to re-run the check?
Screen Shot 2022-10-04 at 18 42 36

**Comment by parthea** [comment]
@cojenco Please could you take a look?

**Comment by cojenco** [comment]
> > Hi @simonbohnen,
> > Please could you sign the CLA? https://cla.developers.google.com/about
> 
> It looks like I did for my work account (SimonBohnenQC). Is there a way to re-run the check? Screen Shot 2022-10-04 at 18 42 36

I force rescanned the CLA and it's passing now. Our team will work on reviewing your PR. Thanks for your contribution!

File: google/resumable_media/requests/download.py [hunk 1]
```diff
@@ -124,7 +124,8 @@ def _write_to_stream(self, response):
                 self._bytes_downloaded += len(chunk)
                 local_checksum_object.update(chunk)
 
-        if expected_checksum is not None:
+        # Don't validate the checksum for partial responses.
+        if expected_checksum is not None and response.status_code != 206:
```
Comment (cojenco) [inline google/resumable_media/requests/download.py:128]: Let's check the status code against `http.client.PARTIAL_CONTENT`

File: google/resumable_media/requests/download.py [hunk 2]
```diff
@@ -308,7 +309,8 @@ def _write_to_stream(self, response):
                 checksum_object.update(chunk)
             response._content_consumed = True
 
-        if expected_checksum is not None:
+        # Don't validate the checksum for partial responses.
+        if expected_checksum is not None and response.status_code != 206:
```
Comment (cojenco) [inline google/resumable_media/requests/download.py:313]: Same for the other places, let's use `http.client.PARTIAL_CONTENT`

**Review by cojenco — CHANGES_REQUESTED** [review]
Thanks @simonbohnen for the fix. I agree we want to skip checksum validations for partial content responses, as noted in our documentation.

In addition to the comments, could you add a unit test for this? Something alongside the tests [here](https://github.com/googleapis/google-resumable-media-python/blob/main/tests/unit/requests/test_download.py#L81). Let me know if you'd like assistance on testing.

**Comment by simonbohnen** [comment]
> Thanks @simonbohnen for the fix. I agree we want to skip checksum validations for partial content responses, as noted in our documentation.
> 
> 
> 
> In addition to the comments, could you add a unit test for this? Something alongside the tests [here](https://github.com/googleapis/google-resumable-media-python/blob/main/tests/unit/requests/test_download.py#L81). Let me know if you'd like assistance on testing. 

I'll add some tests next week👌🏼

**Comment by simonbohnen** [comment]
Can you help me with this error? The commit messages look good to me 🤔
Screen Shot 2022-10-24 at 14 51 21

**Review by cojenco — COMMENTED** [review]
Thanks for adding tests @simonbohnen 

Besides the lint error on unused `ret_val` in the tests, LGTM. Try running `nox -s lint` to check linting.

**Review by cojenco — APPROVED** [review]
LGTM, thanks again for your contribution!

**Pull request merged**